src/test/java/build/buildfarm/instance/memory/MemoryInstanceTest.java

```diff
@@ -0,0 +1,161 @@
+
+  @Mock
+  private ContentAddressableStorage storage;
+
+  @Before
+  public void setUp() throws Exception {
+    MockitoAnnotations.initMocks(this);
+    outstandingOperations = new HashMap<>();
+    MemoryInstanceConfig memoryInstanceConfig = MemoryInstanceConfig.newBuilder()
+        .setListOperationsDefaultPageSize(1024)
+        .setListOperationsMaxPageSize(16384)
+        .setTreeDefaultPageSize(1024)
+        .setTreeMaxPageSize(16384)
+        .setOperationPollTimeout(Duration.newBuilder()
+            .setSeconds(10)
+            .setNanos(0))
+        .setOperationCompletedDelay(Duration.newBuilder()
+            .setSeconds(10)
+            .setNanos(0))
+        .setDefaultActionTimeout(Duration.newBuilder()
+            .setSeconds(600)
+            .setNanos(0))
+        .setMaximumActionTimeout(Duration.newBuilder()
+            .setSeconds(3600)
+            .setNanos(0))
+        .build();
+
+    instance = new MemoryInstance(
+        "memory",
+        new DigestUtil(DigestUtil.HashFunction.SHA256),
+        memoryInstanceConfig,
+        storage,
+        outstandingOperations);
+  }
+
+  @Test
+  public void listOperationsForEmptyOutstanding() {
+    ImmutableList.Builder<Operation> operations = new ImmutableList.Builder<>();
+
+    String nextToken = instance.listOperations(
+        /* pageSize=*/ 1,
+        /* pageToken=*/ "",
+        /* filter=*/ "",
+        operations);
+    assertThat(nextToken).isEqualTo("");
+    assertThat(operations.build()).isEmpty();
+  }
+
+  @Test
+  public void listOperationsForOutstandingOperations() {
+    Operation operation = Operation.newBuilder()
+        .setName("test-operation")
+        .build();
+
+    outstandingOperations.put(operation.getName(), operation);
+
+    ImmutableList.Builder<Operation> operations = new ImmutableList.Builder<>();
+
+    String nextToken = instance.listOperations(
+        /* pageSize=*/ 1,
+        /* pageToken=*/ "",
+        /* filter=*/ "",
+        operations);
+    // we should have reached the end of list
+    assertThat(nextToken).isEqualTo("");
+    assertThat(operations.build()).containsExactly(operation);
+  }
+
+  @Test
+  public void listOperationsLimitsPages() {
+    Operation testOperation1 = Operation.newBuilder()
+        .setName("test-operation1")
+        .build();
+
+    Operation testOperation2 = Operation.newBuilder()
+        .setName("test-operation2")
+        .build();
+
+    outstandingOperations.put(testOperation1.getName(), testOperation1);
+    outstandingOperations.put(testOperation2.getName(), testOperation2);
+
+    ImmutableList.Builder<Operation> operations = new ImmutableList.Builder<>();
+
+    String nextToken = instance.listOperations(
+        /* pageSize=*/ 1,
+        /* pageToken=*/ "",
+        /* filter=*/ "",
+        operations);
+    // we should not be at the end
+    assertThat(nextToken).isNotEqualTo("");
+    assertThat(operations.build().size()).isEqualTo(1);
```

Comment:
It seems you can merge this test case with the next one without making them less clear (you expect to get 1, and then another one).